setup-pyenv.sh

```diff
@@ -37,21 +99,52 @@ export PATH="$PYENV_ROOT/bin:$PATH"
 eval "$(pyenv init -)"
 
 # Make sure the cache directory exists
-PYTHON_BUILD_CACHE_PATH="${PYTHON_BUILD_CACHE_PATH:-$HOME/.pyenv_cache}"
-mkdir -p "$PYTHON_BUILD_CACHE_PATH"
+mkdir -p "$PYENV_CACHE_PATH"
+
+# Try using an already cached PYENV_VERSION. If it fails or is not found,
+# then install from scratch.
+echo "**** Trying to find and use cached python $PYENV_VERSION."
+if ! use_cached_python; then
+  echo "**** Installing python $PYENV_VERSION with pyenv now."
+  if pyenv install "$PYENV_VERSION"; then
+    if mv "$version_pyenv_path" "$PYENV_CACHE_PATH"; then
+      echo "Python was successfully built and moved to cache."
+      echo "**** Trying to find and use cached python $PYENV_VERSION."
+      if ! use_cached_python; then
+        echo "Python version $PYENV_VERSION was apparently successfully built"
+        echo "with pyenv, but, once cached, it could not be verified."
+        output_debugging_info
+        return 1
+      fi
+    else
+      echo "** Warning: Python was succesfully built, but moving to cache"
+      echo "failed. Proceeding anyway without caching."
+    fi
```

Comment:
It's not really clear what failure cases are expected here? I guess it's alright.. this is just the most complicated bit of code and I'm trying to understand what could've gone wrong at each branch.